File: app/src/main/java/com/bailout/stickk/ubi4/utility/BlockingQueueUbi4.java
package com.bailout.stickk.ubi4.utility;

import java .util.ArrayList;

public class BlockingQueueUbi4 {
    ArrayList<Runnable> tasks = new ArrayList<>();
    private boolean canTake = true; // Флаг, разрешающий извлечение задачи
    private long lastAllowTime = 0; // Время последнего события dataReceive

    public synchronized Runnable get() {
        while (tasks.isEmpty() || !canTake) {
            try {
                if (!tasks.isEmpty() && !canTake) {
                    // Проверяем, прошла ли секунда с последнего dataReceive
                    long elapsed = System.currentTimeMillis() - lastAllowTime;
                    if (elapsed >= 1000) {
                        canTake = true; // Автоматическая разблокировка
                    } else {
                        // Ждём оставшееся время до секунды
                        wait(1000 - elapsed);
                    }
                } else {
                    wait();
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        Runnable task = tasks.get(0);
        canTake = false;
        tasks.remove(task);
        return task;
    }

    public synchronized void put(Runnable task, byte[] byteArray) {
        tasks.add(task);
        StringBuilder byteArrayS = new StringBuilder();
        for (byte b : byteArray) {
            byteArrayS.append(" "+b);
        }
        notify();
    }

    public synchronized int size() {
        return tasks.size();
    }

    public synchronized void allowNext() {
        canTake = true;
        lastAllowTime = System.currentTimeMillis(); // Фиксируем время события
        notify(); // Разрешаем извлечение следующей задачи
    }
}

File: shared/src/commonMain/kotlin/com/bailout/stickk/ubi4/utility/BlockingQueueUbi4.kt
package com.bailout.stickk.ubi4.utility

import com.bailout.stickk.ubi4.utility.logging.platformLog
import kotlinx.coroutines.Runnable


class BlockingQueueUbi4 {
    //    private val tasks = mutableListOf<Runnable>()
    private data class QueueEntry(
        val task: Runnable,
        val enqueuedAt: Long,
        val description: String
    )

    private val tasks = mutableListOf<QueueEntry>()
    private var canTake: Boolean = true // Флаг, разрешающий извлечение задачи
    private var lastAllowTime: Long = 0L // Время последнего события (например, dataReceive)
    private val autoUnlockMs = 200L

    fun get(): Runnable {
        // Используем цикл с коротким сном, чтобы избежать busy loop
        while (true) {
            synchronized(this) {
                if (!canTake) {
                    val elapsed = currentTimeMillis() - lastAllowTime
                    if (elapsed >= autoUnlockMs) {
                        canTake = true // Автоматическая разблокировка спустя задержку
                        platformLog(
                            "sendBytesKmm",
                            "BlockingQueueUbi4: авто-разблокировка через ${elapsed}мс, в очереди ${tasks.size}"
                        )
                    }
                }

                if (tasks.isNotEmpty() && canTake) {
                    val entry = tasks.removeAt(0)
                    val waitMs = currentTimeMillis() - entry.enqueuedAt
                    lastAllowTime = currentTimeMillis()
                    canTake = false
                    platformLog(
                        "sendBytesKmm",
                        "BlockingQueueUbi4: берём задачу спустя ${waitMs}мс, в очереди осталось ${tasks.size}. data=${entry.description}"
                    )
                    return entry.task
                }
            }
            sleep(10) // Короткий сон (50 мс) для экономии ресурсов
        }
    }

    fun put(task: Runnable, byteArray: ByteArray) {
        val now = currentTimeMillis()
        synchronized(this) {
            val description = "len=${byteArray.size} hex=${EncodeByteToHex.bytesToHexString(byteArray)}"
            tasks.add(QueueEntry(task = task, enqueuedAt = currentTimeMillis(), description = description))

            val lockedForMs = if (canTake) 0L else (now - lastAllowTime)

            platformLog(
                "sendBytesKmm",
                "BlockingQueueUbi4: put +1, size=${tasks.size}, canTake=$canTake, lockedFor=${lockedForMs}ms, data=$description"
            )
        }
    }

    fun size(): Int {
        return synchronized(this) {
            tasks.size
        }
    }

    fun allowNext(deviceAddress: Int, parameterID: Int,  receiveDataString: String) {
        synchronized(this) {
            platformLog(
                "sendBytesKmm",
                "А тут разрешаем протолкнуть следующую команду allowNext  deviceAddress = $deviceAddress   parameterID = $parameterID   data = $receiveDataString"
            )
            canTake = true
            lastAllowTime = currentTimeMillis() // Фиксируем время события
        }
    }
}
